### ISSUE TITLE
[question] Optional client cert check

### ISSUE THREAD

**Issue by niteshagrawalgmail** [istio/istio#31984]
Hello,

I have a question related to securing the ingress gateway.
As per the documentation, I understand that for mtls the Gateway resource configuration - spec.servers.tls.mode should be MUTUAL.
This is associated with a 'host' which we mention in the spec.servers.hosts
once this MUTUAL configuration is enabled, Is it possible to send calls to the host without a client certificate?
I know that I can have another host configuration, for which the tls mode could be SIMPLE.
But my use case is to use the same host for client cert auth and basic auth. Does istio support some kind of OPTIONAL check for certificate?

Regards,
Nitesh

**Comment by howardjohn** [comment]
There must be some way to do this in Envoy, but looking at its docs I don't see how this is possible...

**Comment by budanm** [comment]
Hi @howardjohn  
The tls context, associated with filter chain of envoy proxy configuration has validation_context (which contains trusted_ca which verifies the client certificate) and a require_client_certificate field

When require_client_certificate is true, if the client mandatorily needs to send the certificate else the calls will fail.

When require_client_certificate is false, the calls pass even if the client does not send the certificate. However, if client certificate is passed , then mtls will happen. This is like optional client certificate authentication

**Comment by howardjohn** [comment]
@budanm my understanding is the client will not send the certificate unless envoy specifically requests it in the TLS handshake though. This is why Go has a bunch of options:
```go
	// NoClientCert indicates that no client certificate should be requested
	// during the handshake, and if any certificates are sent they will not
	// be verified.
	NoClientCert ClientAuthType = iota
	// RequestClientCert indicates that a client certificate should be requested
	// during the handshake, but does not require that the client send any
	// certificates.
	RequestClientCert
	// RequireAnyClientCert indicates that a client certificate should be requested
	// during the handshake, and that at least one certificate is required to be
	// sent by the client, but that certificate is not required to be valid.
	RequireAnyClientCert
	// VerifyClientCertIfGiven indicates that a client certificate should be requested
	// during the handshake, but does not require that the client sends a
	// certificate. If the client does send a certificate it is required to be
	// valid.
	VerifyClientCertIfGiven
	// RequireAndVerifyClientCert indicates that a client certificate should be requested
	// during the handshake, and that at least one valid certificate is required
	// to be sent by the client.
	RequireAndVerifyClientCert
```

**Comment by sonnyhcl** [comment]
Hi @howardjohn , we are facing unexpected behavior for `tlsmode=SIMPLE` and `tlsmode=MUTUAL` when we upgrade from 1.9.9 to 1.10.0.
Below is our `gateway.yaml`. We are using `tlsmode=SIMPLE` since our gateway will accept both token auth and client auth. Also we set `caCertificates` to validate client cert if has.
But when we upgrade from 1.9.9 to 1.10.0, we found our client cert is not honored by gateway from `curl-1.10.0-simple.log`, which is different from `curl-1.9.9-simple.log`.

By dumping ingressgateway proxy-config from ingressgateway (`k exec -ti istio-ingressgateway-7f677b4fdf-4jp2l -n istio-system -- curl http://localhost:15000/config_dump > istio-ingressgateway-7f677b4fdf-4jp2l.pc.json`), I found 1.9.9 has `require_client_certificate=false` and have set `validation_context` from `caCertficates` we set, while 1.10.0 only has `require_client_certificate=false` and don't have any `validation_context` config. 

I think it's related to this commit #31391, which is first released in istio 1.10.0. 
After going through the conversations in #31391 and this issue, I think this should be a regression. `tlsmode=SIMPLE` should also have ability to optional check client cert, which should set `validation_context` if user specify `caCertificates`.

```yaml
# gateway.yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: ingressgateway
  namespace: default
spec:
  selector:
    istio: ingressgateway
  servers:
    - port:
        number: 443
        name: https
        protocol: HTTPS
      tls:
        mode: SIMPLE
        minProtocolVersion: TLSV1_2
        serverCertificate: /etc/ingressgateway-https-certs/tls.crt
        privateKey: /etc/ingressgateway-https-certs/tls.key
        caCertificates: /etc/ca-certs/ca.pem
```
```log
#curl-1.10.0-simple.log
$ curl --cert-type p12 --cert client-cert.pfx https://xxx -vvv
*   Trying 20.109.145.255...
* TCP_NODELAY set
* Connected to xxx (20.109.145.255) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Unknown (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Client hello (1):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
```
```
#curl-1.9.9-simple.log
$ curl --cert-type p12 --cert client-cert.pfx https://xxx -vvv
*   Trying 20.109.145.255...
* TCP_NODELAY set
* Connected to xxx port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Unknown (8):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Client hello (1):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, CERT verify (15):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
```
